BaragonService/src/main/java/com/hubspot/baragon/service/worker/BaragonRequestWorker.java

```diff
@@ -242,56 +249,124 @@ public void run() {
-
-            if (!maybeState.isPresent()) {
-              LOG.warn(String.format("%s does not have a request status!", requestId));
-              continue;
-            }
+      while (added < configuration.getWorkerConfiguration().getMaxRequestsPerPoll() && !queuedRequestIds.isEmpty()) {
+        Map<String, List<QueuedRequestId>> requestsGroupedByService = queuedRequestIds.stream().collect(Collectors.groupingBy(QueuedRequestId::getServiceId));
 
-            final Optional<BaragonRequest> maybeRequest = requestManager.getRequest(requestId);
+        ArrayList<QueuedRequestId> nonServiceChanges = new ArrayList<>();
+        ArrayList<QueuedRequestId> serviceChanges = new ArrayList<>();
 
-            if (!maybeRequest.isPresent()) {
-              LOG.warn(String.format("%s does not have a request object!", requestId));
-              continue;
+        for (Map.Entry<String, List<QueuedRequestId>> requestsForService : requestsGroupedByService.entrySet()) {
+          for (QueuedRequestId request : requestsForService.getValue()) {
+            if (added >= configuration.getWorkerConfiguration().getMaxRequestsPerPoll()) {
+              break;
             }
 
-            queuedRequestsWithState.add(new QueuedRequestWithState(queuedRequestId, maybeRequest.get(), maybeState.get()));
-            handledServices.add(queuedRequestId.getServiceId());
-            added++;
+            // Grab as many non-service-change BaragonRequests as we can.
+            if (requestManager.getRequest(request.getRequestId()).transform(stateDatastore::isServiceUnchanged).or(false)) {
+              nonServiceChanges.add(request);
+              added++;
+            } else {
+              // Once we hit a BaragonRequest that specifies BaragonService changes, stop collecting requests for this service.
+              serviceChanges.add(request);
+              added++;
+              break;
+            }
           }
-          if (added >= configuration.getWorkerConfiguration().getMaxRequestsPerPoll()) {
-            break;
+        }
+
+        // Now take the list of non-service-change requests,
+        // hydrate them with state,
+        // and sort them such that the quicker noValidate / noReload requests come first.
+        List<QueuedRequestWithState> hydratedNonServiceChanges = nonServiceChanges.stream()
+            .map(this::hydrateQueuedRequestWithState)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .sorted(queuedRequestComparator())
+            .collect(Collectors.toList());
+
+        // Then send them off.
+        LOG.debug("Processing {} BaragonRequests which don't modify a BaragonService", nonServiceChanges.size());
+        results.putAll(handleQueuedRequests(hydratedNonServiceChanges));
+
+        // Now send off the service change requests.
+        for (QueuedRequestId serviceChange : serviceChanges) {
```

Comment:
It looks like `serviceChanges` is always a 0 or 1 item list. Maybe make it an optional to make the logic more readable? At first it looked like this was bypassing other logic like checking `added` until I realized not much gets added to that initial `serviceChanges` list